### ISSUE TITLE
Campaign Finished before all emails Sent

### ISSUE THREAD

**Issue by malvibid** [knadh/listmonk#1802]
- listmonk: v3.0.0
 - OS: Linux

I need to send a Newsletter to all 1,450 Subscribers in my Newsletter list. I started the campaign, but the campaign finished after sending only 578 / 1,450. Upon investigating I have found there is a 454 Throttling Failure. 

![image](https://github.com/knadh/listmonk/assets/100029192/95b013aa-b364-4d0e-989a-592b9d79cd31)

**Extract of Logs:**
```
2024/03/30 07:05:32start processing campaign (March Newsletter)
2024/03/30 07:05:38error sending message in campaign March Newsletter: subscriber 243: 454 Throttling failure: Maximum sending rate exceeded.
2024/03/30 07:05:38error sending message in campaign March Newsletter: subscriber 268: 454 Throttling failure: Maximum sending rate exceeded.
2024/03/30 07:05:38error sending message in campaign March Newsletter: subscriber 269: 454 Throttling failure: Maximum sending rate exceeded.
2024/03/30 07:05:38error sending message in campaign  March Newsletter: subscriber 252: 454 Throttling failure: Maximum sending rate exceeded.
...
2024/03/30 07:05:55error sending message in campaign March Newsletter: subscriber 1447: 454 Throttling failure: Maximum sending rate exceeded.
2024/03/30 07:05:55campaign (March Newsletter) finished
2024/03/30 07:05:55error sending message 'Finished: March Newsletter': 454 Throttling failure: Maximum sending rate exceeded.

```

My performance settings were the default. 

Concurrency: 10
Message Rate: 10
Batch Size: 10000
Maximum Error Threshold: 1000

I have now adjusted the settings as follows.

![image](https://github.com/knadh/listmonk/assets/100029192/52293c4b-75b5-471d-b6e7-030c2e2b0388)

1.) Would these settings prevent the campaign from Finishing before all the emails are sent? And Pause the campaign and allow me to investigate incase there is an error?

2.) There's ~870 subscribers in this list that didn't receive the email. I want to create a second list to resend the emails to these subscribers only. But how do I retrieve this list? The logs only show me, what I assume is the Subscribers ID. I have extracted these ID's using Excel functions. How can I use these ID's to create a new list - `Newsletter-2`?

**Comment by MaximilianKohler** [comment]
>Would these settings prevent the campaign from Finishing before all the emails are sent?

You have to find out from your SMTP provider. For example, Amazon SES starts you out at around 12 per second I think, and you can gradually ask them for increases.

2.) Yeah, I have that same problem. Some info in my comments here https://github.com/knadh/listmonk/issues/1762

**Comment by malvibid** [comment]
Thanks for clarifying. I resolved email throttling issues in Listmonk by adjusting the message rate to 12 messages per second and the sliding window limit to 720 messages per minute. 

To identify subscribers who hadn't received the email, I followed these steps:

1. Saved logs before restarting Listmonk, as restarts clear logs.
2. Extracted missing subscriber IDs from the logs using Excel’s `TEXTSPLIT` function.
3. Exported subscriber details from the Listmonk database to a CSV file:
    ```
    psql -U listmonk -d listmonk
    \copy (SELECT * FROM subscribers) TO '/path/to/your/subscribers.csv' WITH CSV HEADER;
    ```
    This was necessary because Listmonk's GUI doesn't support ID searches or include the database ID in exports.
4. Used `XLOOKUP` in Excel to match the extracted IDs with subscriber details from the CSV.
5. Imported the identified subscribers into a new List in Listmonk and resent a copy of the email campaign to them.

**Enhancement Suggestion:** The workaround was effective, but it's a roundabout way of managing newsletter lists and errors, leaving me with extra cleanup. It would greatly benefit users if Listmonk simplified this process within the app itself, by making it easier to identify undelivered emails to avoid potential spamming with duplicate messages.

In any case, thank you for building and open-sourcing such an amazing tool.

**Comment by knadh** [comment]
> The workaround was effective, but it's a roundabout way of managing newsletter lists and errors, leaving me with extra cleanup. It would greatly benefit users if Listmonk simplified this process within the app itself, by making it easier to identify undelivered emails to avoid potential spamming with duplicate messages.

Hi @malvibid. You can setup bounce tracking to accurately account for and automatically disable problematic e-mails.

**Comment by malvibid** [comment]
Thank you for you reply. I have set up bounce processing, and have no known issues there. But what I meant was identifying undelivered emails if a campaign finishes before all subscribers of a list are sent an email.

In my case, only 578 / 1,450 subscribers were sent an email and the campaign Finished. So, 872 people never got the email. It'd useful to know who these 872 people were so that I can try to send them an email again. The logs give me the subscriber ID, but its not practical for me to check these one-by-one for 872 people, but it is really the only option available in the app at the moment.

Additionally, my original list is called `Newsletter` with 1450 subs, and I had to create a new List with these 872 people called - `Newsletter-2`.  This makes list management slightly tricky. Ideally, I do not want to create a new list, but maybe have just a temporary sub-list/segment of the same list `Newsletter` that I can send the emails to.

**Comment by MaximilianKohler** [comment]
Yeah, I don't think bounce tracking is related to this at all. And some of us aren't lucky enough to get subscriber IDs in the logs. https://github.com/knadh/listmonk/issues/1762

Even with malvibid's helpful directions it would take me ages to figure out how to do all that.

It would be great to have something like https://sesdashboard.com/ integrated into listmonk.

**Comment by knadh** [comment]
@malvibid I agree. In an ideal scenario, failed campaign e-mails have to be logged in a table separately and there has to be a way to retry/resend a campaign just to those IDs. I've thought about it from time to time (as an extension to segmentation/ephemeral lists) but haven't had any easy enough approaches. In the last version, the concurrent messaging core was completely re-written and that should make achieving this simpler. I'll consider this for the next version.

Re-opening this issue for tracking.

**Comment by dwrolvink** [comment]
For us, we use an internal email relay. We got a load of these errors:
```
2024/05/14 13:58:41 manager.go:485: error sending message in campaign Kubernetes 1.28 - Upgradeschema: subscriber 32576: timed out waiting for free conn in pool
```

When sending a test email in the SMTP settings pane, it took about 20 seconds. With the SMTP settings being default, could it be that Listmonk is sending more emails per second than that are resolved per second, "filling up the pool"? (Not sure how this works technically).

We switched to a secondary relay and this caused the test message to be sent immediately again. Hoping that this will fix this for us now momentarily.

**Comment by mjp0** [comment]
We just ran into this same issue and now apparently I'm going to spend hours doing manual cross-referencing from Postgres to figure out who got the email and who didn't. Absolutely wonderful, this is exactly the type of stuff I wanted to be doing today.

In summary, this "high performance" email software can't even send emails properly to whole list without crashing and burning in the process if there's throttling errors, and there's no recovery. Maybe you should mention this "limitation" in the README, so anybody who is looking for a mailing list manager with an actual email list knows to stay away.

**Comment by knadh** [comment]
The bug that caused subscribers to be skipped in the campaign has been fixed: https://github.com/knadh/listmonk/issues/1931#issuecomment-2371792556

**Comment by latticepublishing** [comment]
I ran into the same issue. Amazon SES defaults the send rate (outside of sandbox) at 14 sends/sec, but Listmonk defaults the send rate to 10 sends/sec with 10 concurrent threads, so effectively a 100 sends/sec rate. So this is definitely my fault that I was throttled because of the settings oversight, but I suspect other email service providers have similar throttling settings.

To add to this problem, the subscribers that did not receive the campaign emails do not appear to be tracked in Listmonk's DB, and the logs found in the Listmonk web UI (Settings -> Logs) appear to be transient so I didn't know which subscribers did not receive the campaign email. I needed to extract the logs from the Docker container which appear to be persisted permanently by default, thankfully.

I haven't worked on this codebase, nor do I have the time to, so I'm definitely not complaining. However, it would be a marked improvement to also track "delivered" emails per campaign and per subscriber.

**Comment by kousu** [comment]
Just checking but @latticepublishing what version are you running? This was fixed in [4.1.0](https://github.com/knadh/listmonk/releases/tag/v4.1.0) ("https://github.com/knadh/listmonk/commit/cea65c009d3b6675ae5a92bf507f233c502db541 Fix and refactor subscriber batch fetching in campaign processing.") and it's been stable for me since then, though I don't send campaigns _that_ often, and there is at least [one report](https://github.com/knadh/listmonk/issues/2278#issuecomment-2738417721) of lossy mails in 4.1.0. There have also been refactorings in 5.0.

**Comment by latticepublishing** [comment]
> Just checking but [@latticepublishing](https://github.com/latticepublishing) what version are you running? This was fixed in [4.1.0](https://github.com/knadh/listmonk/releases/tag/v4.1.0) ("[cea65c0](https://github.com/knadh/listmonk/commit/cea65c009d3b6675ae5a92bf507f233c502db541) Fix and refactor subscriber batch fetching in campaign processing.") and it's been stable for me since then, though I don't send campaigns _that_ often, and there is at least [one report](https://github.com/knadh/listmonk/issues/2278#issuecomment-2738417721) of lossy mails in 4.1.0. There have also been refactorings in 5.0.

Hey @kousu , thanks for taking interest! I'm running 5.0.0 and in my first campaign I had these settings:

Campaign name: 806 - national
- Concurrency: 10
- Message Rate: 10
- Batch Size: 10000
- Maximum Error Threshold: 1000

I've since adjusted my settings to these below and ran a new campaign to the remaining subscribers that received the throttling errors I was able to extract from the Docker container logs:

Campaign name: 806 - national patch
- Concurrency: 1
- Message Rate: 10
- Batch Size: 10000
- Maximum Error Threshold: 1000

The 2nd campaign appears to have completed properly, but the total subscribers associated with that campaign was less than the successful send count in the original campaign so perhaps that could be obfuscating the same issue.

Also, for what it's worth, the bounce figures in the summary stats are not reflecting what I'm seeing in my Amazon SES console. Perhaps a few regression bugs were introduced in 5.0.0.

<img width="1213" height="611" alt="Image" src="https://github.com/user-attachments/assets/6f960052-fc36-42d1-950c-06acb4a8b6b0" />

**Comment by kousu** [comment]
Ah okay that's too bad, but good to know I still need to pay attention!

I thought my situation was particularly extreme because I'm using a verrrrrry cheap SMTP plan. I assumed people using SES would be in the same league as, say, MailGun, and would never see a problem.

btw I've never seen the bounce figures be accurate. I haven't dug in that hard though -- I assume I just don't have my bounce mailbox credentials set up properly, or something. If it's not just me maybe it's easy to misconfigure?

**Comment by MaximilianKohler** [comment]
>Also, for what it's worth, the bounce figures in the summary stats are not reflecting what I'm seeing in my Amazon SES console. 

SES only records hard bounces & complaints. Listmonk records soft bounces and more.

**Comment by latticepublishing** [comment]
> > Also, for what it's worth, the bounce figures in the summary stats are not reflecting what I'm seeing in my Amazon SES console.
> 
> SES only records hard bounces & complaints. Listmonk records soft bounces and more.

Hey @MaximilianKohler , thanks for the information. I'm a little new to email marketing, but in my Amazon SES console I see that they track "transient bounces". I was under the impression that these are considered "soft bounces". Am I understanding this incorrectly? 

<img width="719" height="551" alt="Image" src="https://github.com/user-attachments/assets/2fffbcb2-9fe4-4f5b-89c1-5d5ecd9c60b2" />

But also, either way, all of my production campaigns show 0 bounces despite the stats in Amazon SES showing non-zero values. It's a little odd because I ran a test campaign sending to the Amazon complaint and hard bounce simulator accounts, and those stats register correctly.

**Comment by MaximilianKohler** [comment]
Oh, yes it looks like they're tracking soft bounces there. Is that CloudWatch? I was thinking of the main SES page where it logs bounced & complained email addresses.

I haven't checked that UI for a while. Your screenshot looks much more user-friendly than I remember it.

I recall there being multiple ways of setting things up with SES. For example, I think if you opt to use their open and/or click tracking, it would conflict with the listmonk one. I can't remember exactly. And since it looks like you're using SES to track opens & clicks that might be one issue.

>Also, for what it's worth, the bounce figures in the summary stats are not reflecting what I'm seeing in my Amazon SES console. Perhaps a few regression bugs were introduced in 5.0.0.

I re-read your issue. Were the bounce figures working correctly for you in the past, and then stopped working correctly with v5.0? Or have you only ever sent 3 total campaigns? I see your bounce test was in Nov 2024, can you run that again and see what happens?

You're using these settings right? https://listmonk.app/docs/bounces/#amazon-simple-email-service-ses

**Comment by latticepublishing** [comment]
@MaximilianKohler thanks for the suggestions. I revisited my setup, per your suggestion, and the issue was my fault. I switched our Listmonk instance to send from a different domain a few months ago and did not reconfigure the notification settings in Amazon SES and SNS. Sorry for wasting your time!

However, I did have a question with regards to the campaign summary stats. These stats don't line up with the metrics that I find in the analytics screen. Am I missing something else obvious? I rescanned the documentation, but I couldn't find anything. Thanks in advance for any help!

Summary Stats
<img width="1203" height="151" alt="Image" src="https://github.com/user-attachments/assets/8514034d-d7a9-4a42-9d92-86774c9e2bae" />

Analytics Stats

<img width="1239" height="794" alt="Image" src="https://github.com/user-attachments/assets/d220f92a-efb3-4a13-8d95-27845e2e070c" />
